pkg/server/conn_stmt.go

```diff
@@ -331,97 +336,143 @@ func (cc *clientConn) executePreparedStmtAndWriteResult(ctx context.Context, stm
 	// we should hold the ResultSet in PreparedStatement for next stmt_fetch, and only send back ColumnInfo.
 	// Tell the client cursor exists in server by setting proper serverStatus.
 	if useCursor {
-		crs := resultset.WrapWithCursor(rs)
-
-		cc.initResultEncoder(ctx)
-		defer cc.rsEncoder.Clean()
-		// fetch all results of the resultSet, and stored them locally, so that the future `FETCH` command can read
-		// the rows directly to avoid running executor and accessing shared params/variables in the session
-		// NOTE: chunk should not be allocated from the connection allocator, which will reset after executing this command
-		// but the rows are still needed in the following FETCH command.
-
-		// create the row container to manage spill
-		// this `rowContainer` will be released when the statement (or the connection) is closed.
-		rowContainer := chunk.NewRowContainer(crs.FieldTypes(), vars.MaxChunkSize)
-		rowContainer.GetMemTracker().AttachTo(vars.MemTracker)
-		rowContainer.GetMemTracker().SetLabel(memory.LabelForCursorFetch)
-		rowContainer.GetDiskTracker().AttachTo(vars.DiskTracker)
-		rowContainer.GetDiskTracker().SetLabel(memory.LabelForCursorFetch)
-		if variable.EnableTmpStorageOnOOM.Load() {
-			failpoint.Inject("testCursorFetchSpill", func(val failpoint.Value) {
-				if val, ok := val.(bool); val && ok {
-					actionSpill := rowContainer.ActionSpillForTest()
-					defer actionSpill.WaitForTest()
-				}
-			})
-			action := memory.NewActionWithPriority(rowContainer.ActionSpill(), memory.DefCursorFetchSpillPriority)
-			vars.MemTracker.FallbackOldAndSetNewAction(action)
-		}
-		defer func() {
-			if err != nil {
-				rowContainer.GetMemTracker().Detach()
-				rowContainer.GetDiskTracker().Detach()
-				errCloseRowContainer := rowContainer.Close()
-				if errCloseRowContainer != nil {
-					logutil.Logger(ctx).Error("Fail to close rowContainer in error handler. May cause resource leak",
-						zap.NamedError("original-error", err), zap.NamedError("close-error", errCloseRowContainer))
-				}
-			}
-		}()
+		lazy, err = cc.executeWithCursor(ctx, stmt, rs)
+		return false, err
+	}
+	retryable, err := cc.writeResultSet(ctx, rs, true, cc.ctx.Status(), 0)
+	if err != nil {
+		return retryable, errors.Annotate(err, cc.preparedStmt2String(uint32(stmt.ID())))
+	}
+	return false, nil
+}
 
-		for {
-			chk := crs.NewChunk(nil)
+func (cc *clientConn) executeWithCursor(ctx context.Context, stmt PreparedStatement, rs resultset.ResultSet) (lazy bool, err error) {
+	vars := (&cc.ctx).GetSessionVars()
+	if vars.EnableLazyCursorFetch {
```

Comment:
Most of the following logic in this function is copied from the previous deleted lines. The only difference is this branch.